src/Linters/NamespacePrivateLinter.hack

```diff
@@ -0,0 +1,148 @@
+/*
+ *  Copyright (c) 2017-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the MIT license found in the
+ *  LICENSE file in the root directory of this source tree.
+ *
+ */
+
+namespace Facebook\HHAST;
+use namespace HH\Lib\{Vec, C, Str};
+final class NamespacePrivateLinter extends ASTLinter {
+	const type TContext = Script;
+	const type TNode = Script;
+
+	<<__Override>>
+	public function getLintErrorForNode(this::TContext $context, this::TNode $node): ?ASTLintError {
+		// Process all namespaces, use statements present in the file.
+		$namespaces_details = $node->getNamespaces();
+
+		// Go through namespace details and its uses to check for potential violations of private namespace boundaries.
+		foreach ($namespaces_details as $namespace_detail_shape) {
+			$current_namespace = $namespace_detail_shape['name'] ?? '';
+			$namespace_uses = Shapes::toArray($namespace_detail_shape['uses']);
+
+			foreach ($namespace_uses as $namespace_use_dict) {
+				foreach ($namespace_use_dict as $namespace_path) {
+					if (!$this->isPrivateNamespacePathAllowed($namespace_path, $current_namespace)) {
+						return new ASTLintError(
+							$this,
+							'Artifacts belonging to a namespace are private to the parent namespace',
+							$node,
+						);
+					}
+				}
+			}
+		}
+
+		$traversed_qualified_names = varray[];
+		foreach ($node->getDescendantsOfType(QualifiedName::class) as $qualified_name_token) {
+			$name_token_key = '';
+			if ($qualified_name_token->hasParts()) {
+				$name_token_key = $qualified_name_token->getDescendantsOfType(NameToken::class)
+					|> Vec\map($$, $qualified_name_token ==> $qualified_name_token->getText())
+					|> Str\join($$, '\\');
+				if (qualified_name_is_fully_qualified($qualified_name_token)) {
+					$name_token_key = '\\'.$name_token_key;
+				}
+			}
+
+			// If a qualified token has already been traversed or has no parts, then skip it.
+      		if (Str\is_empty($name_token_key) || C\contains($traversed_qualified_names, $name_token_key)) {
+				continue;
+			}
+			// Add the current qualified name token to the traversed list.
+			$traversed_qualified_names[] = $name_token_key;
+
+			// get the namespace for each qualified name token that we encounter.
+			foreach ($namespaces_details as $ns) {
+				if ($ns['children']->isAncestorOf($qualified_name_token)) {
+					$current_namespace = $ns['name'] ?? '';
+					$fully_qualified_name_for_current_token = '';
+
+					$parent_node = $context->getParentOfDescendant($qualified_name_token);
+					if ($parent_node is ScopeResolutionExpression || $parent_node is NameExpression) {
+						$fully_qualified_name_for_current_token = $this->resolve_scope(
+							$name_token_key,
+							$ns['uses']['namespaces'],
+							$current_namespace
+						);
+					} else if ($parent_node is FunctionCallExpression) {
+						$fully_qualified_name_for_current_token = resolve_function(
```

Comment:
`resolve_function()` and `resolve_type()` also use `get_uses_directly_in_scope()` under the hood, same as used by `getNamespaces()` - not sure if there is an efficient way of doing this using the existing code.

Reply:
It's probably worth us moving that to scripts/nodes and caching it... but that'll eat more memory :(